### ISSUE TITLE
Property hasOwnProperty never defined on type Object

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#866]
With new type inference, getting the following incorrect warning:
```
src/test/TestViewerApp.js:416: WARNING - Property hasOwnProperty never defined on {}
    if (!c.hasOwnProperty(p)) continue;  // skip inherited properties
         ^
```
on this method:
```javascript
/** Adds tests from the given class to the menu of available tests.
Tests are static methods that take a parameter of type {ContactSim}
and return {undefined}.
The test added must end with "_setup";  the menu name is generated from
the test method name by changing all underscores to blanks, and removing
the "_setup" suffix.
* @param {!Object} c the class whose tests will be added to the menu of available tests.
* @private
*/
TestViewerApp.prototype.addTestsFrom = function(c) {
  //window.console.log("addTestsFrom "+c);
  for (var p in c) {
    if (!c.hasOwnProperty(p)) continue;  // skip inherited properties
    if (typeof c[p] !== "function") continue;  // skip non-functions
    if (!p.match(/.*_setup$/)) continue;  // skip non-setup functions
    var nm = p.replace(/_setup$/, "").replace(/_/g, " ");
    //window.console.log("testName="+nm);
    this.testNames.push(nm);
    this.tests.push(c[p]);
  }
};
```

**Comment by dimvar** [comment]
I just tried this example and I can't reproduce the issue. It must have been fixed at some earlier commit.

    function f(/** !Object */ x) {
      return x.hasOwnProperty('asdf');
    }

**Comment by myphysicslab** [comment]
This is still a problem in Version: v20150505-58-g39bb337.  I tried your example and I get the error.  Perhaps I'm using different compiler flags?

```javascript
function f(/** !Object */ x) {
  return x.hasOwnProperty('asdf');
};
```
results in:
```
src/test/TestViewerApp.js:458: WARNING - Property hasOwnProperty never defined on {}
  return x.hasOwnProperty('asdf');
         ^
```
Here is the compile command:
```
java -jar ../javascript/closure-compiler/build/compiler.jar --new_type_inf
--compilation_level=SIMPLE --closure_entry_point=myphysicslab.test.TestViewerApp
--js=src --js=../javascript/closure-library --only_closure_dependencies=true
--generate_exports --jscomp_error=checkTypes --jscomp_error=missingProperties
--jscomp_error=missingProvide --jscomp_error=missingRequire
--jscomp_warning=accessControls --jscomp_warning=ambiguousFunctionDecl
--jscomp_warning=checkVars --jscomp_warning=const --jscomp_warning=constantProperty
--jscomp_warning=fileoverviewTags --jscomp_warning=globalThis
--jscomp_warning=invalidCasts --jscomp_warning=misplacedTypeAnnotation
--jscomp_warning=missingReturn --jscomp_warning=newCheckTypes
--jscomp_warning=strictModuleDepCheck --jscomp_warning=suspiciousCode
--jscomp_warning=typeInvalidation --jscomp_warning=undefinedNames
--jscomp_warning=undefinedVars --jscomp_warning=unknownDefines
--jscomp_warning=uselessCode --jscomp_warning=visibility --extra_annotation_name=alias
--warning_level=VERBOSE --define=myphysicslab.lab.util.UtilityCore.ADVANCED=false
--define=goog.DEBUG=true '--define=goog.LOCALE='\''en'\''' --language_in=ECMASCRIPT5
```

**Comment by dimvar** [comment]
I tried it in the debugger and I can't see the error. See:
http://closure-compiler-debugger.appspot.com/home#input0%3Dfunction%2520f(%252F**%2520!Object%2520*%252F%2520x)%2520%257B%250A%2520%2520return%2520x.hasOwnProperty('asdf')%253B%250A%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1

(There are 4 warnings because NTI can't understand the es6 externs well, and 1 warning in html5.js which is a bug in that file and should be fixed.)

Can you try using the compiler at head instead of the latest release?

**Comment by myphysicslab** [comment]
I just did a pull and recompiled the compiler.
```
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20150505-63-g4e69889
Built on: 2015/05/24 23:41
```
Still getting the error.  Just noticed it's happening in closure library also:
```
../javascript/closure-library/closure/goog/disposable/disposable.js:110: WARNING - Property 
hasOwnProperty never defined on {}
    if (goog.Disposable.instances_.hasOwnProperty(id)) {
        ^

../javascript/closure-library/closure/goog/disposable/disposable.js:185: WARNING - Property 
hasOwnProperty never defined on {}
          !goog.Disposable.instances_.hasOwnProperty(uid)) {
           ^

src/test/TestViewerApp.js:439: WARNING - Property hasOwnProperty never defined on {}
    if (!c.hasOwnProperty(p)) continue;  // skip inherited properties
         ^

src/test/TestViewerApp.js:450: WARNING - Property hasOwnProperty never defined on {}
  return x.hasOwnProperty('asdf');
         ^
```

I've used git a lot but only locally, I've not worked with a remote much.  `git status` was telling me I was up to date, but then I did a pull and got a bunch of changes... confusing.

```
23:31:04 ~/Documents/Programming/javascript/closure-compiler]$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working directory clean
[23:31:08 ~/Documents/Programming/javascript/closure-compiler]$ git diff origin/master
[23:33:12 ~/Documents/Programming/javascript/closure-compiler]$ git diff HEAD origin/master
[23:33:51 ~/Documents/Programming/javascript/closure-compiler]$ git pull
remote: Counting objects: 56, done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 56 (delta 10), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (56/56), done.
From https://github.com/google/closure-compiler
   39bb337..4e69889  master     -> origin/master
Updating 39bb337..4e69889
Fast-forward
 contrib/externs/chrome_extensions.js               | 222 +-------------------
 contrib/externs/jasmine-2.0.js                     |   4 +
 externs/chrome.js                                  | 224 +++++++++++++++++++++
 .../javascript/jscomp/DefaultPassConfig.java       |  21 +-
 src/com/google/javascript/jscomp/NodeUtil.java     |   4 +-
 src/com/google/javascript/jscomp/PolymerPass.java  |  31 ++-
 .../jscomp/RemoveUnusedClassProperties.java        |  25 ++-
 .../google/javascript/jscomp/ScopedAliases.java    |  82 +++++---
 .../google/javascript/jscomp/PolymerPassTest.java  |  97 ++++-----
 .../jscomp/RemoveUnusedClassPropertiesTest.java    |  16 ++
 .../javascript/jscomp/ScopedAliasesTest.java       |  46 ++++-
 11 files changed, 447 insertions(+), 325 deletions(-)
```

**Comment by dimvar** [comment]
I usually use the compiler version that is internal to Google. I just tried to reproduce this with the open-source compiler and failed. So I'm fairly certain that this is an issue with your setup. Here are the steps I took.

<code>git pull origin master</code>
To get the latest commits. You'll need to do git clone if you don't have the repository locally already.

<code>ant jar</code>
To create the compiler jar

<code>java -jar build/compiler.jar --new_type_inf --js=tmp.js</code>
To run the program, where tmp.js contains the small example I posted earlier.

**Comment by dimvar** [comment]
Try removing <code>--jscomp_error=missingProperties</code>.

**Comment by dimvar** [comment]
Unrelated.

Also, when NTI is on, we turn off some other passes because they are covered by NTI.
https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CompilerOptionsPreprocessor.java#L68

So, the following options you're using have no effect.

    --jscomp_error=checkTypes
    --jscomp_warning=globalThis
    --jscomp_warning=missingReturn

**Comment by myphysicslab** [comment]
Any of these four options cause the "Property hasOwnProperty never defined" warning to appear:
```
--jscomp_warning=newCheckTypes
--warning_level=VERBOSE
--jscomp_error=missingProperties
--jscomp_error=checkTypes
```

For example:
```
$ java -jar build/compiler.jar --new_type_inf --js=tmp.js  --jscomp_error=checkTypes

tmp.js:2: ERROR - Property hasOwnProperty never defined on {}
  return x.hasOwnProperty('asdf');
         ^

1 error(s), 7 warning(s)
```

The file tmp.js is:
```javascript
function f(/** !Object */ x) {
  return x.hasOwnProperty('asdf');
}
```

In compiling my large program, I have to remove *ALL  4* of these options to have the `hasOwnProperty` warning not appear.  The total number of warnings then goes down from 630 to 575 (I haven't tried to figure out what other warnings are disappearing).

I'm pretty sure I'm building the compiler correctly using ant.  Here is the version:
```
+ java -jar ../javascript/closure-compiler/build/compiler.jar --version
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20150505-63-g4e69889
Built on: 2015/05/24 23:41
```

Should I not compile with those 4 options when using NTI?  I assume I want to keep those options when not using NTI.  (I cannot use NTI for my regular work because there are still too many warnings to be practical).

**Comment by dimvar** [comment]
Are you using custom externs or the default externs?

**Comment by myphysicslab** [comment]
Default externs.  Here is a typical compile command (note the troublesome 4 options have been removed):

```
+ java -jar ../javascript/closure-compiler/build/compiler.jar --new_type_inf
--compilation_level=SIMPLE --closure_entry_point=myphysicslab.test.TestViewerApp
--js=src --js=../javascript/closure-library --only_closure_dependencies=true
--generate_exports --jscomp_error=missingProvide --jscomp_error=missingRequire
--jscomp_warning=accessControls --jscomp_warning=ambiguousFunctionDecl
--jscomp_warning=checkVars --jscomp_warning=const
--jscomp_warning=constantProperty --jscomp_warning=fileoverviewTags
--jscomp_warning=globalThis --jscomp_warning=invalidCasts
--jscomp_warning=misplacedTypeAnnotation --jscomp_warning=missingReturn
--jscomp_warning=strictModuleDepCheck --jscomp_warning=suspiciousCode
--jscomp_warning=typeInvalidation --jscomp_warning=undefinedNames
--jscomp_warning=undefinedVars --jscomp_warning=unknownDefines
--jscomp_warning=uselessCode --jscomp_warning=visibility
--extra_annotation_name=alias
--define=myphysicslab.lab.util.UtilityCore.ADVANCED=false --define=goog.DEBUG=true
'--define=goog.LOCALE='\''en'\''' --language_in=ECMASCRIPT5
```

**Comment by dimvar** [comment]
The issue is that the open-source compiler runner doesn't use the DefaultPassConfig, so it's not properly canceling the other passes when NTI is enabled. Will fix.

**Comment by dimvar** [comment]
Wrong diagnosis in my last comment. Anyway, I can reproduce the bug consistently now.

**Comment by dimvar** [comment]
Note to self: this is probably related to the fact that object types can have a null nominal type instead of inheriting from Object. That fix will likely fix this too.

**Comment by myphysicslab** [comment]
Confirmed fixed in Version: v20150609-23-g19613b6. I am able to compile with all these (possibly useless) compiler options:
```
--new_type_inf \
--jscomp_warning=newCheckTypes
--warning_level=VERBOSE
--jscomp_error=missingProperties
--jscomp_error=checkTypes
```
and no longer see the "Property hasOwnProperty never defined" warning.
